### DISCUSSION TITLE
How do I get a security token?

### DISCUSSION THREAD

**Discussion by Verhoeckx** [gitpod-io/openvscode-server#249] [Q&A] [answered]
Hello OpenVSCode developers,

Somewhere in the coming weeks I want to try out OpenVSCode and in preparation I read the documenation.

I do not completely understand the text below

> Visit the URL printed in your terminal. Just going to port 3000 won't work, because VS Code requires that you provide a uniquely generated security token to prevent unauthorized access.

Okay, I understand this: I need some kind of security token.

> Since OpenVSCode Server v1.62, you must access the Web UI using a connection token generated by the server.

Okay, the same as above.

> If that's not what you're looking for, and instead want to generate these tokens by yourself, start the server with --connectionToken YOUR_TOKEN to force your secret to be used. 

I don't care if the token is generated or that I have to create one myself as long as I can run the application.

**Questions that arises:**

1. How/when are the security tokens generated? How do I get one?
2. If I can use my own token, how should this token look like? Can it be anything? Every word, password of passphrase?

> If you want your development environment to be even more secure, create a plaintext file with the desired token as its contents and provide it to the server with --connection-secret YOUR_SECRET_FILE.

This is clear.

**Comment by filiptronicek** [comment, marked as answer]
Hi there, @Verhoeckx, I am glad that you asked these questions, seems like the docs perhaps need a bit of a refresh. 

1. The tokens are generated as a random UUID (if you want to explore the format, you can see some examples and the RegExp for them [here](https://onlinestringtools.com/generate-string-from-regex?regex=%5B0-9a-f%5D%7B8%7D-%5B0-9a-f%5D%7B4%7D-%5B0-9a-f%5D%7B4%7D-%5B0-9a-f%5D%7B4%7D-%5B0-9a-f%5D%7B12%7D%24&results=10)), they are printed as a query parameter inside of the terminal when starting the server 
![image](https://user-images.githubusercontent.com/29888641/143826594-4bdb7f0d-f30f-4d3a-be70-b50714170560.png)

2. If you want to use your own token, as you mentioned you provide it with `--connection-token=YOUR_DESIRED_TOKEN`. This can be **any string**, just make sure to put it inside `''` if it has spaces in it.

**Reply by Verhoeckx** [reply]
Hello @filiptronicek,

Thank your very much for your answers! It's now clear to me where the token can be found: in the terminal of the server! I didn't understand that from the documentation but now I do.

When I have finished setting up the virtual machine: I will definitely try out OpenVScode!

Regards - Verhoeckx

**Reply by santiagobasulto** [reply]
Is it possible to specify it as an environment variable?

**Comment by asabi** [comment]
When I try to run the container like so:

```docker run --rm -it --init --connection-token='secret token' -p 3000:3000 -v "$(pwd):/home/workspace:cached" gitpod/openvscode-server```

or

```docker run --rm -it --init --connection-secret ='/path/to/secret/file' -p 3000:3000 -v "$(pwd):/home/workspace:cached" gitpod/openvscode-server```

I get

```unknown flag: --connection-token``` / ```unknown flag: --connection-secret```

What am I doing wrong?

**Reply by akosyakov** [reply]
Have you tried to add it after `gitpod/openvscode-server`?

**Reply by asabi** [reply]
Awesome, thank you!
